### PR TITLE
Enable builds for ARM64

### PR DESCRIPTION
**Pull request by artpdr**
- Generates artifacts for AMD64 and ARM64 by default
- ARM64 compilation is done using emulation, hence it's slow but doesn't require a new toolchain and system root files that would take lots of space in the repo and possibly have legal issues
- There's the possibility of computing only for AMD64 which is way faster and most likely what's desirable for testing quick developments

### CODE REVIEW

File: make.sh [hunk 1]
```diff
@@ -67,30 +68,49 @@ echo "build/ is not up-to-date and no build caches found. Proceeding with build.
...
+echo_stage "Building LightGBM CI docker image replicas..."
+bash docker/make_docker_image.sh amd64
+
+if [[ "$ARCH_BUILD" != "single" ]]; then
+  bash docker/make_docker_image.sh arm64
+fi
 
-echo_stage "Building LightGBM CI docker image replica..."
-bash docker/make_docker_image.sh
+echo_stage "Launching container AMD64..."
```
Comment (AlbertoEAF) [inline make.sh]:
```suggestion
echo_stage "Launching AMD64 container..."
```

File: make.sh [hunk 2]
```diff
@@ -67,30 +68,49 @@ echo "build/ is not up-to-date and no build caches found. Proceeding with build.
...
-container=$(docker run -e LIGHTGBM_REPO_URL -t -d lightgbm-ci-build-env)
-docker cp docker/lightgbm-ci-build-env/make_lightgbm.sh $container:/lightgbm
-echo_bold "Running container: $container"
+echo_stage "Building LightGBM AMD64 $LIGHTGBM_VERSION..."
+docker container exec $container_amd64 bash make_lightgbm.sh "$LIGHTGBM_VERSION"
 
-echo_stage "Building LightGBM $LIGHTGBM_VERSION..."
-docker container exec $container bash make_lightgbm.sh "$LIGHTGBM_VERSION"
+if [[ "$ARCH_BUILD" != "single" ]]; then
+  echo_stage "Launching container ARM64..."
```
Comment (AlbertoEAF) [inline make.sh]:
```suggestion
  echo_stage "Launching ARM64 container..."
```

File: make_patch.sh [hunk 3]
```diff
@@ -37,8 +37,8 @@ echo_bold "Compiling with existing CMake settings at build_dir=${BUILD_LGBM_PATC
 CWD="$(pwd)"
 cd "$BUILD_LGBM_PATCH_FROM"
 make -j4
-find . -name "*.so" -exec cp {} "${CWD}/build" \;
+find . -name "*.so" -exec cp {} "${CWD}/build/amd64" \;
 cp lightgbmlib.jar "${CWD}/build"
-cp $(ldd "${CWD}/build/lib_lightgbm.so" | awk '/libgomp/{print $3}') "${CWD}/build/libgomp.so.1.0.0"
+cp $(ldd "${CWD}/build/amd64/lib_lightgbm.so" | awk '/libgomp/{print $3}') "${CWD}/build/amd64/libgomp.so.1.0.0"
```
Comment (AlbertoEAF) [inline make_patch.sh:50]: We need to add a comment to this file (maybe here and in the header) and a note on the README (in the respective section) with a caveat that this currently only applies for AMD64 parts of the build.

File: make.sh [hunk 4]
```diff
@@ -67,30 +68,49 @@ echo "build/ is not up-to-date and no build caches found. Proceeding with build.
...
+echo_stage "Launching container AMD64..."
+container_amd64=$(docker run -e LIGHTGBM_REPO_URL -t -d lightgbm-ci-build-env-amd64)
+docker cp docker/make_lightgbm.sh $container_amd64:/lightgbm
+echo_bold "Running container AMD64: $container_amd64"
 
-echo_stage "Launching container..."
-container=$(docker run -e LIGHTGBM_REPO_URL -t -d lightgbm-ci-build-env)
-docker cp docker/lightgbm-ci-build-env/make_lightgbm.sh $container:/lightgbm
-echo_bold "Running container: $container"
+echo_stage "Building LightGBM AMD64 $LIGHTGBM_VERSION..."
```
Comment (AlbertoEAF) [inline make.sh]:
```suggestion
echo_stage "Building AMD64 LightGBM $LIGHTGBM_VERSION..."
```

File: make.sh [hunk 5]
```diff
@@ -67,30 +68,49 @@ echo "build/ is not up-to-date and no build caches found. Proceeding with build.
...
+if [[ "$ARCH_BUILD" != "single" ]]; then
+  bash docker/make_docker_image.sh arm64
+fi
 
-echo_stage "Building LightGBM CI docker image replica..."
-bash docker/make_docker_image.sh
+echo_stage "Launching container AMD64..."
+container_amd64=$(docker run -e LIGHTGBM_REPO_URL -t -d lightgbm-ci-build-env-amd64)
+docker cp docker/make_lightgbm.sh $container_amd64:/lightgbm
+echo_bold "Running container AMD64: $container_amd64"
```
Comment (AlbertoEAF) [inline make.sh]:
```suggestion
echo_bold "Running AMD64 container: $container_amd64"
```

File: make.sh [hunk 6]
```diff
@@ -67,30 +68,49 @@ echo "build/ is not up-to-date and no build caches found. Proceeding with build.
...
+echo_stage "Building LightGBM AMD64 $LIGHTGBM_VERSION..."
+docker container exec $container_amd64 bash make_lightgbm.sh "$LIGHTGBM_VERSION"
 
-echo_stage "Building LightGBM $LIGHTGBM_VERSION..."
-docker container exec $container bash make_lightgbm.sh "$LIGHTGBM_VERSION"
+if [[ "$ARCH_BUILD" != "single" ]]; then
+  echo_stage "Launching container ARM64..."
+  container_arm64=$(docker run --platform=arm64 -e LIGHTGBM_REPO_URL -t -d lightgbm-ci-build-env-arm64)
+  docker cp docker/make_lightgbm.sh $container_arm64:/lightgbm
+  echo_stage "Building LightGBM ARM64 $LIGHTGBM_VERSION..."
```
Comment (AlbertoEAF) [inline make.sh]:
```suggestion
  echo_stage "Building ARM64 LightGBM $LIGHTGBM_VERSION..."
```

File: docker/make_docker_image.sh [hunk 7]
```diff
@@ -19,4 +19,18 @@
 set -e
 
 cd docker
-docker build -t lightgbm-ci-build-env lightgbm-ci-build-env
+
+case "$1" in
+	arm64)
+		echo "Building docker image for ARM64"
+    docker build -t lightgbm-ci-build-env-arm64 --platform=arm64 lightgbm-ci-build-env-arm64
```
Comment (AlbertoEAF) [inline docker/make_docker_image.sh]: broken indentation, same below, use spaces

File: make.sh [hunk 8]
```diff
@@ -26,6 +26,8 @@ export LIGHTGBM_REPO_URL="${LIGHTGBM_REPO_URL:-https://github.com/microsoft/Ligh
 
 LIGHTGBM_VERSION=$([[ -z "$1" ]] && echo "master" || echo "$1")
 PACKAGE_VERSION="$2"
+# If $ARCH_BUILD == single we generate only artifacts for AMD64
+ARCH_BUILD="${ARCH_BUILD:-multi}"
```
Comment (AlbertoEAF) [inline make.sh]: move the new lines to a new isolated paragraph after line 38, so the lgbm version configuration is all together as before.

**Review by AlbertoEAF — CHANGES_REQUESTED** [review]
We could refactor this a bit but as we've talked let's ship this ASAP. Trivial style comments only

File: docker/lightgbm-ci-build-env-arm64/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        git \
+        tar \
+        gnupg-curl \
+ && rm -rf /var/lib/apt/lists/* \
+ && rm -rf /etc/apt/sources.list.d/*
+
+# Install CMake
+RUN curl -OL https://github.com/Kitware/CMake/releases/download/v3.27.4/cmake-3.27.4-linux-aarch64.sh \
+ && curl -OL https://github.com/Kitware/CMake/releases/download/v3.27.4/cmake-3.27.4-SHA-256.txt \
+ && curl -OL https://github.com/Kitware/CMake/releases/download/v3.27.4/cmake-3.27.4-SHA-256.txt.asc \
```
Comment (AlbertoEAF) [inline docker/lightgbm-ci-build-env-arm64/Dockerfile:27]: This is not very safe if there is any future tampering. 

Put this `.asc` file directly in in the `docker/lightgbm-ci-build-env-arm64/` folder and copy it to the image and verify the checksum through that file.

**Comment by AlbertoEAF** [comment]
Also, we need to add some README detailing how to run these docker ARM builds locally. It's not that trivial and could be useful to other developers.

File: docker/make_docker_image.sh (hunk 7 above)
Comment (artpdr) [reply docker/make_docker_image.sh]: done

File: make.sh (hunk 8 above)
Comment (artpdr) [reply make.sh]: done

File: make_patch.sh (hunk 3 above)
Comment (artpdr) [reply make_patch.sh:50]: Added a comment to the header and readme.

**Review by AlbertoEAF — CHANGES_REQUESTED** [review]

File: make.sh (hunk 8 above)
Comment (AlbertoEAF) [reply make.sh]: did you push? looks the same

File: docker/lightgbm-ci-build-env-arm64/Dockerfile (hunk 9 above)
Comment (artpdr) [reply docker/lightgbm-ci-build-env-arm64/Dockerfile:27]: We verify this asc below using gpg

File: make.sh [hunk 10]
```diff
@@ -67,30 +69,49 @@ echo "build/ is not up-to-date and no build caches found. Proceeding with build.
...
-container=$(docker run -e LIGHTGBM_REPO_URL -t -d lightgbm-ci-build-env)
-docker cp docker/lightgbm-ci-build-env/make_lightgbm.sh $container:/lightgbm
-echo_bold "Running container: $container"
+echo_stage "Building AMD64 LightGBM $LIGHTGBM_VERSION..."
+docker container exec $container_amd64 bash make_lightgbm.sh "$LIGHTGBM_VERSION"
 
-echo_stage "Building LightGBM $LIGHTGBM_VERSION..."
-docker container exec $container bash make_lightgbm.sh "$LIGHTGBM_VERSION"
+if [[ "$ARCH_BUILD" != "single" ]]; then
+    echo_stage "Launching container ARM64..."
```
Comment (AlbertoEAF) [inline make.sh]:
```suggestion
    echo_stage "Launching ARM64 container..."
```

File: make.sh [hunk 11]
```diff
@@ -67,30 +69,49 @@ echo "build/ is not up-to-date and no build caches found. Proceeding with build.
...
+    container_arm64=$(docker run --platform=arm64 -e LIGHTGBM_REPO_URL -t -d lightgbm-ci-build-env-arm64)
+    docker cp docker/make_lightgbm.sh $container_arm64:/lightgbm
+    echo_stage "Building LightGBM ARM64 $LIGHTGBM_VERSION..."
```
Comment (AlbertoEAF) [inline make.sh]:
```suggestion
    echo_stage "Building ARM64 LightGBM $LIGHTGBM_VERSION..."
```

**Review by AlbertoEAF — CHANGES_REQUESTED** [review]

**Review by AlbertoEAF — APPROVED** [review]

**Pull request merged**